File: src/FragmentRenderer.ts
import { Container, Sprite, Texture } from "pixi.js";
import SpellCircleRenderer from "./SpellCircleRenderer";
import Fragment from "./fragment/Fragment";
import Pattern from "./fragment/Pattern";
import SpellPart from "./fragment/SpellPart";
import PatternGlyph from "./fragment/PatternGlyph";

export default interface FragmentRenderer<T extends Fragment> {
    render(
        fragment: T,
        container: Container,
        x: number,
        y: number,
        size: number,
        alpha: number,
        textures: Map<string, Texture>,
        delegator: SpellCircleRenderer
    ): void;

    renderRedrawDots(): boolean;
}

const fragmentRenderers: Map<string, FragmentRenderer<Fragment>> = new Map();



export { fragmentRenderers };

File: site/src/jsMain/kotlin/me/maplesyrum/tricksterStudio/components/webComponents/spellElement/FragmentRenderer.kt
package me.maplesyrum.tricksterStudio.components.webComponents.spellElement

import me.maplesyrum.tricksterStudio.Identifier
import me.maplesyrum.tricksterStudio.external.pixi.Container
import me.maplesyrum.tricksterStudio.external.pixi.Texture
import me.maplesyrum.tricksterStudio.spell.fragment.Fragment

interface FragmentRenderer<T : Fragment> {
    fun render(
        fragment: T,
        container: Container,
        x: Double,
        y: Double,
        size: Double,
        alpha: Double,
        textures: Map<String, Texture>,
        delegator: SpellCircleRenderer
    )

    fun renderRedrawDots(): Boolean
}

val fragmentRenderers: MutableMap<Identifier, FragmentRenderer<Fragment>> = mutableMapOf()

